rfcs/20210307-tfx-api-artifact-compat.md

```diff
@@ -0,0 +1,600 @@
+        their environment, the scaffolded pipeline can still function.
+
+#### Examples, Notebooks and Tutorials {#examples-notebooks-and-tutorials}
+
+Examples, Notebooks (Colab) and Tutorials are **excluded** from the
+compatibility compliance. These are created mostly for demonstrating how to use
+TFX in accordance with best practices, therefore often subject to a particular
+version of TFX.
+
+These examples should use the stable public API of `tfx` as much as possible.
+Usage of experimental or deprecated APIs in these examples should be explicitly
+called out, and the examples should pin their dependency version of `tfx` down
+to a narrow range as much as possible in these rare cases.
+
+#### New Orchestrator {#new-orchestrator}
+
+The new orchestrator is in early stages of development and we don’t think it’ll
+be production-ready for 1.0 timeline, thus it will enter 1.0 as `experimental`.
+
+
+#### Intermediate Representation (IR) and Container entrypoint {#ir-entrypoint}
+
+Both the intermediate representataion (IR) and container entrypoint of official
+TFX images are under active development and **excluded** from the stability /
+compatiblity scope of this doc.
+
+#### Portability Layer to Other Runners (aka Orchestrators) {#portability-layer-to-other-runner-orchestrators}
+
+This portability layer in tfx refers to various DagRunners which runs TFX
+pipelines on various external orchestrators, the _portable_ module and the
+abstraction between CLI and Dag runner. These libraries will comply with
+backward compatibility.
+
+Notice that several Runners are experimental by their nature, or slated to be
+deprecated. Specifically, we plan to support the following runners with backward
+compatibility:
+
+*   LocalDagRunner
+*   KubeflowV2DagRunner
+
+The following Dag runners will remain experimental:
+
+*   AirflowDagRunner
+    *   Due to low utilization of the AirflowDagRunner, we want to avoid calling
+        it stable. Support will continue to graduate it into production as utilization evolves.
+
+The following dag runners are slated to be deprecated and therefore won’t
+comply:
+
+*   BeamDagRunner: this will be replaced by the LocaDagRunner
+*   KubeflowDagRunner: this dag runner implementation based on KFP v1 API will
+    be replaced by the V2 dag runner above.
+
+#### FrontEnd {#frontend}
+
+Similar to Orchestrator, for the upcoming TFX::OSS frontend (first version as
+plugin to Tensorboard), no backward/forward compatibility is guaranteed.
+
+#### Command Line Interface (CLI) {#command-line-interface-cli}
+
+Command groups to fulfill key CUJs will be backward compatible, notably:
+
+*   _pipeline_
+*   _run_
+
+_The template_ command group will remain experimental.
+
+### ML-Metadata (MLMD) {#ml-metadata-mlmd}
+
+We are proposing that MLMD project enters the same 1.0 readiness and
+compatibility as TFX. The rest of this section will discuss various aspects of
+MLMD.
```

Comment:
nit: how about context types? Will those be stable (currently they are privately defined - see [this](https://github.com/tensorflow/tfx/blob/master/tfx/orchestration/metadata.py#L65) ). Mostly asking as TFX Addons MLMD Client may need to rely on some those to retrieve them

Reply:
Good question. Based on my discussion with 
@ruoyu90 this is highly coupled with IR, so we'd view them as the same thing  and probably graduate them in a lock step in the future.